SQL: Add text formatting support for multivalue

bpintea opened the pull request:
This adds the text formatting for multivalue fields (available through the ARRAY function),
 that is:
- CSV and TSV exports;
- TXT formatting.

bpintea committed:
Add text formatting support for multivalue

This adds the text formatting for multivalue doc fields, that is:
- CSV and TSV exports
- TXT formatting

matriv commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/rest/RestSqlTestCase.java:
```diff
@@ -1055,6 +1055,16 @@ private void executeQueryWithNextPage(String format, String expectedHeader, Stri
     }
 
+    public void testMultiValueQueryText() throws IOException {
+        index("{\"text\":[" + toJson("one") + "," + toJson("two, three") + "," + toJson("\"four\"") + "], \"number\" : [1, [2, 3], 4] }");
```
Why do you use `toJson` only for the values but not also for the field names?

bpintea replied:
C&p inertia. Thx, updated it.

matriv commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plugin/TextFormatTests.java:
```diff
@@ -164,6 +313,24 @@ private static SqlQueryResponse emptyData() {
+                }
+
+                String typeName = fromJava(o).typeName();
+                headers.add(new ColumnInfo("index", typeName + "_column", typeName));
```
I don't remember our take here, shouldn't the type name reveal that it's an array? something like `byte_array` or `_array` (postgres format) ?

bpintea replied:
You mean the 3rd argument should be `typeName + "_array"`? I've updated it anyways, to be clearer, but it's not used in these tests (the formatters don't look at the type, only content).

matriv replied:
Thx, yep, I get, the columnInfo is not actually asserted, but it's clearer with the `_array` suffix. Thx!

matriv reviewed (commented):
Looks good, left 2 comments, a minor one and a question regarding the data type.

bpintea committed:
Merge branch 'feat/sql-multivalue' into feat/multivalue_text

bpintea committed:
Adress review comments

- extend `toJson` use in one test;
- add + `"_array"` in column def in one test.

bpintea committed:
Apply spotless formatting

- which is pretty ugly in this case.

matriv reviewed (approved):
LGTM. Thx!

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plugin/TextFormatTests.java:
```diff
@@ -152,6 +156,151 @@ public void testInvalidCsvDelims() {
+
+    public void testCsvMultiValueIntegers() {
+        String text = CSV.format(req(), withData(singletonList(asList(
+            asList((byte)1, (byte)2), asList((short)3, (short)4), asList(5, 6), asList(7L, 8L)
```
Nitty picky: whitespace between the casting and the casted value.

astefan reviewed (approved):
LGTM

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

costin commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/StringUtils.java:
```diff
@@ -166,6 +167,31 @@ public static String toString(Object value, SqlVersion sqlVersion) {
+        // multivalue
+        if (value instanceof Collection) {
+            final StringBuilder sb = new StringBuilder();
+            sb.append('[');
+            Collection<?> values = (Collection<?>) value;
+            values.forEach(x -> {
+                // quote strings and `\`-escape the `"` character inside them
+                if (x instanceof String) { // TODO: IPs should ideally not be quoted.
+                    sb.append('"');
+                    sb.append(((String) x).replace("\"", "\\\""));
+                    sb.append('"');
+                } else if (x == null) {
+                    sb.append("NULL");
+                } else {
+                    sb.append(toString(x, sqlVersion));
+                }
+                sb.append(',');
+            });
+            if (sb.length() > 1) { // strip trailing comma
+                sb.deleteCharAt(sb.length() - 1);
+            }
+            sb.append(']');
+            return sb.toString();
+        }
```
Better to use `StringJoiner` since it's part of the JDK and handles the exact case of prefix/suffix plus delimiter.

bpintea replied:
Nice, thanks, applied.

costin reviewed (approved):
LGTM

bpintea committed:
Address review comments

- white-space style;
- s/StringBuilder/StringJoiner/

matriv commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/StringUtils.java:
```diff
@@ -169,27 +170,19 @@ public static String toString(Object value, SqlVersion sqlVersion) {
         if (value instanceof Collection) {
-            final StringBuilder sb = new StringBuilder();
-            sb.append('[');
+            final StringJoiner sj = new StringJoiner(",", "[", "]");
```
+1 nice!

bpintea committed:
Merge branch 'feat/sql-multivalue' into feat/multivalue_text

bpintea committed:
Merge branch 'feat/sql-multivalue' into feat/multivalue_text

palesz commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/StringUtils.java:
```diff
@@ -166,6 +168,23 @@ public static String toString(Object value, SqlVersion sqlVersion) {
+                if (x instanceof String) { // TODO: IPs should ideally not be quoted.
+                    sj.add('"' + ((String) x).replace("\"", "\\\"") + '"');
+                } else if (x == null) {
+                    sj.add("NULL");
```
Nitpicky: we return uppercase `NULL`, but [here](https://github.com/elastic/elasticsearch/pull/68606/files#diff-c28edb24b62ae39893a3d5f5591e035cfe910e4d042faeec230a34d589da1b74R96) we return lowercase `null` for simple null objects.

bpintea replied:
Sorry, GH only updated the page with your comment after I hit merge.
The link in the comment doesn't seem to resolve correctly, but I assume you refer to the first `if` in the function? 
I followed Posgresql's capitalised formatting for lists. Generally (and debatable) nulls are capitalised in textual formats, so I'd be in favour for updating the listing of simple null objects.

As extra context:
The `NULL` would appear "alone" in CSVs and TSVs -- a simple null object wouldn't be output at all in these formats (they'd be represented by two consecutive commas or tabs).
In CLI, the `NULL` would also appear to the letter in arrays output and the CLI chooses how to represent JSON nulls by itself (which is currently small lettered).
In `txt` formatting output the difference would be visible and "uncontrollable".

palesz reviewed (approved):
LGTM
